src/sources/exec.rs

```diff
@@ -0,0 +1,886 @@
+use crate::async_read::VecAsyncReadExt;
+use crate::config::{DataType, SourceContext};
+use crate::event::LogEvent;
+use crate::internal_events::{ExecCommandExecuted, ExecTimeout};
+use crate::{
+    config::{log_schema, SourceConfig, SourceDescription},
+    event::Event,
+    internal_events::{ExecEventReceived, ExecFailed},
+    shutdown::ShutdownSignal,
+    Pipeline,
+};
+use bytes::{Bytes, BytesMut};
+use chrono::Utc;
+use futures::{FutureExt, SinkExt, StreamExt};
+use serde::{Deserialize, Serialize};
+use snafu::Snafu;
+use std::io;
+use std::io::{Error, ErrorKind};
+use std::path::PathBuf;
+use std::process::ExitStatus;
+use tokio::io::{AsyncRead, BufReader};
+use tokio::process::Command;
+use tokio::sync::mpsc::{channel, Sender};
+use tokio::time::{self, sleep, Duration, Instant};
+use tokio_stream::wrappers::IntervalStream;
+use tokio_util::codec::{Decoder, FramedRead, LinesCodec};
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(default, deny_unknown_fields)]
+pub struct ExecConfig {
+    pub mode: Mode,
+    pub scheduled: Option<ScheduledConfig>,
+    pub streaming: Option<StreamingConfig>,
+    pub command: Vec<String>,
+    pub working_directory: Option<PathBuf>,
+    #[serde(default = "default_include_stderr")]
+    pub include_stderr: bool,
+    #[serde(default = "default_events_per_line")]
+    pub event_per_line: bool,
+    #[serde(default = "default_maximum_buffer_size")]
+    pub maximum_buffer_size_bytes: usize,
+}
+
+// TODO: Would be nice to combine the scheduled and streaming config with the mode enum once
+//       this serde ticket has been addressed (https://github.com/serde-rs/serde/issues/2013)
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(rename_all = "snake_case", deny_unknown_fields)]
+pub enum Mode {
+    Scheduled,
+    Streaming,
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(rename_all = "snake_case", deny_unknown_fields)]
```

Comment:
Do we need `rename_all = "snake_case"` here and below? The options are already snake cased.

Reply:
Your right! I have removed it